File: hw5/InversionCount.py
def __main__():
    instances = int(input())

    for i in range(instances):
        inversions = input().split(" ")
        inversions = list((int(i) for i in inversions)) #convert strings to int in list
        temp = [0]*len(inversions) #creating a temp list with same size as list "inversions"
        #[0]*len(x) is a common Python idiom for creating a new list with a specific length and default value (i.e: 0)
        # In the case above, it is creating a new list that has the same length as "inversions" list and every element in the list is init to 0

        print(countInversions(inversions, temp, 0, len(inversions) - 1))
 

def countInversions(array, temp, leftStart, rightEnd):
    
    if leftStart >= rightEnd:
        return 0
    
    midpoint = (leftStart + rightEnd) // 2
    count = 0
    count += countInversions(array, temp, leftStart, midpoint) #front half recursive call
    count += countInversions(array, temp, midpoint + 1, rightEnd) #back half recursive call
    count += merge(array, temp, leftStart, rightEnd) #merge and count array halves

    return count

def merge(array, temp, leftStart, rightEnd):
    leftEnd = (leftStart + rightEnd) // 2 #midpoint
    rightStart = leftEnd + 1 #midpoint + 1
    size = rightEnd - leftStart + 1

    left = leftStart
    right = rightStart
    index = leftStart
    inversions = 0

    while (left <= leftEnd and right <= rightEnd):
        if array[left] <= array[right]:
            temp[index] = array[left]
            left += 1
        else: 
            temp[index] = array[right]
            right += 1
            inversions += leftEnd - left + 1
        index += 1

    array[leftStart:rightEnd+1] = temp[leftStart:rightEnd+1]

    return inversions
        
if __name__ == "__main__":
    __main__()

File: hw5/InversionCount.java
import java.util.Scanner;
import java.util.Arrays;

public class InversionCount {
    public static void main(String[] args) {

        Scanner scnr = new Scanner(System.in);
        int instances = scnr.nextInt(); //getting number of instances from user
        scnr.nextLine();

        for (int i = 0; i < instances; i++){
            long listCount = scnr.nextInt(); //getting number of nodes for current instance
            scnr.nextLine();

            String[] list = scnr.nextLine().split(" ");
            int[] arr = new int[list.length];

            for (int k = 0; k < list.length; k++) { 
                arr[k] = Integer.parseInt(list[k]);
            }

            // System.out.println(countSort(arr, 0, arr.length - 1));
            System.out.println(countInversions(arr));
 
        }
        scnr.close();
    }

    public static long countInversions(int[] arr){
        return countSort(arr, new int[arr.length], 0, arr.length - 1);
    }

    public static long countSort(int[] array, int[] temp, int leftStart, int rightEnd) {
        
        if (leftStart >= rightEnd) {
            return 0;
        }

        int midpoint = (leftStart + rightEnd) / 2;

        long inversions = countSort(array, temp, leftStart, midpoint);
        inversions += countSort(array, temp, midpoint + 1, rightEnd);
        inversions += merge(array, temp, leftStart, rightEnd);   

        return inversions;
    }

    public static long merge (int[] array, int[] temp, int leftStart, int rightEnd) {
        int leftEnd = (leftStart + rightEnd) / 2; // the ending index for the left pointer
        int rightStart = leftEnd + 1; //the start index for the right pointer
        int size = rightEnd - leftStart + 1; //size of the current array

        int left = leftStart;
        int right = rightStart;
        int index = leftStart;
        long inversions = 0;

        while (left <= leftEnd && right <= rightEnd) {
            if (array[left] <= array[right]) {
                temp[index] = array[left];
                left++;
            } else {
                temp[index] = array[right];
                right++;
                inversions += leftEnd - left + 1;
            }
            index++;
        }

        System.arraycopy(array, left, temp, index, leftEnd - left + 1);
        System.arraycopy(array, right, temp, index, rightEnd - right + 1);
        System.arraycopy(temp, leftStart, array, leftStart, size);

        return inversions;
    }
}